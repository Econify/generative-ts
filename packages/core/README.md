# @generative-ts/core

Core functionality for generative-ts, see [main README](https://github.com/Econify/generative-ts/blob/main/README.md)
