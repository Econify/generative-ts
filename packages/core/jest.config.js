const { pathsToModuleNameMapper } = require('ts-jest');
const { compilerOptions } = require('../../tsconfig.base.json');

module.exports = {
  displayName: 'core',
  testEnvironment: 'node',
  rootDir: '../../',
  roots: ['<rootDir>/packages/core/src'],
  transform: {
    '^.+\\.ts$': 'ts-jest'
  },
  testMatch: ['<rootDir>/packages/core/src/**/*.spec.ts'],
  moduleFileExtensions: ['ts', 'js', 'json', 'node'],
  moduleNameMapper: pathsToModuleNameMapper(compilerOptions.paths, { prefix: '<rootDir>/' })
};
