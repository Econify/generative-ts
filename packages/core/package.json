{
  "name": "@generative-ts/core",
  "version": "0.1.0-alpha.5",
  "description": "Core functionalities for generative-ts",
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "exports": {
    "default": "./dist/index.cjs",
    "import": "./dist/index.mjs",
    "types": "./dist/index.d.ts"
  },
  "files": [
    "dist",
    "README.md"
  ],
  "scripts": {
    "build": "npm run clean && npm run build:rollup",
    "build:rollup": "rollup -c",
    "check": "npm run typecheck && npm run lint && npm run test",
    "clean": "rm -rf dist && rm -rf .rollup.cache && rm -f *.tsbuildinfo",
    "lint": "eslint src/",
    "lint:fix": "eslint --fix src/",
    "prepare": "npm run build",
    "prepublishOnly": "npm run check",
    "test": "jest",
    "test:cov": "jest --coverage",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "aws4": "^1.12.0",
    "ejs": "^3.1.10",
    "fp-ts": "^2.16.5",
    "io-ts": "^2.2.21",
    "tslib": "^2.6.2"
  },
  "devDependencies": {
    "@types/ejs": "^3.1.5"
  },
  "publishConfig": {
    "access": "public"
  },
  "license": "ISC",
  "repository": {
    "url": "Econify/generative-ts",
    "directory": "packages/core"
  },
  "author": "John M Naglick",
  "homepage": "https://github.com/Econify/generative-ts#readme",
  "bugs": {
    "url": "https://github.com/Econify/generative-ts/issues"
  },
  "keywords": [
    "llm",
    "openai",
    "gpt",
    "meta",
    "llama",
    "mistral",
    "cohere",
    "bedrock",
    "groq",
    "huggingface",
    "ts",
    "generative",
    "typescript"
  ]
}
