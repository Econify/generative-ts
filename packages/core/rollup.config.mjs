import { baseConfig } from '../../rollup.config.base.mjs';

// import pkg from './package.json' assert { type: 'json' };
const pkg = { name: '@generative-ts/core' };

export default [
  {
    ...baseConfig,
    input: 'src/index.ts',
    output: [
      {
        file: 'dist/index.cjs',
        format: 'cjs',
        sourcemap: true
      },
      {
        file: 'dist/index.mjs',
        format: 'esm',
        sourcemap: true
      }
    ],
    plugins: [
      ...baseConfig.plugins,
      {
        name: 'finished',
        buildStart() {
          console.log(`\x1b[32mBundling ${pkg.name}...\x1b[0m`);
        },
        writeBundle({ format }) {
          console.log(`\x1b[32mFinished \x1b[1m${pkg.name}\x1b[0m (${format})`);
        }
      }
    ],
    external: [
      ...baseConfig.external,
      'aws4',
      'ejs',
      'fp-ts/lib/Either.js',
      'io-ts',
    ]
  }
];
