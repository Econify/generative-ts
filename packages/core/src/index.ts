/*
 * Public API - APIs
 */
export {
  Ai21Jurassic2Api,
  Ai21Jurassic2Options,
  Ai21Jurassic2Response,
  Ai21Jurassic2Template,
} from "./apis/ai21";
export {
  AmazonTitanTextApi,
  AmazonTitanTextOptions,
  AmazonTitanTextResponse,
  AmazonTitanTextTemplate,
} from "./apis/amazon";
export {
  CohereChatApi,
  CohereChatOptions,
  CohereChatResponse,
  CohereChatTemplate,
  CohereGenerateApi,
  CohereGenerateOptions,
  CohereGenerateResponse,
  CohereGenerateTemplate,
} from "./apis/cohere";
export {
  GoogleGeminiApi,
  GoogleGeminiOptions,
  GoogleGeminiResponse,
  GoogleGeminiTemplate,
} from "./apis/google";
export {
  HfConversationalTaskApi,
  HfConversationalTaskOptions,
  HfConversationalTaskResponse,
  HfConversationalTaskTemplate,
  HfInferenceApiOptions,
  HfTextGenerationTaskApi,
  HfTextGenerationTaskOptions,
  HfTextGenerationTaskResponse,
  HfTextGenerationTaskTemplate,
} from "./apis/huggingface";
export {
  LlamaResponse,
  Llama2ChatApi,
  Llama2ChatOptions,
  Llama2ChatTemplate,
  Llama3ChatApi,
  Llama3ChatOptions,
  Llama3ChatTemplate,
} from "./apis/meta";
export {
  MistralAiApi,
  MistralAiOptions,
  MistralAiResponse,
  MistralAiTemplate,
  MistralBedrockApi,
  MistralBedrockOptions,
  MistralBedrockResponse,
  MistralBedrockTemplate,
} from "./apis/mistral";
export {
  OpenAiChatApi,
  OpenAiChatOptions,
  OpenAiChatResponse,
  OpenAiChatTemplate,
} from "./apis/openai";

/*
 * Public API - Providers
 */
export {
  BaseModelProvider,
  HttpModelProvider,
  createAwsBedrockModelProvider,
  createCohereModelProvider,
  createGroqModelProvider,
  createHuggingfaceInferenceModelProvider,
  createLmStudioModelProvider,
  createMistralModelProvider,
  createOpenAiChatModelProvider,
  AwsAuthConfig,
  CohereAuthConfig,
  GroqAuthConfig,
  HuggingfaceAuthConfig,
  MistralAuthConfig,
  OpenAiAuthConfig,
} from "./providers";

/*
 * Public API - Utils
 */
export { EjsTemplate } from "./utils";

/*
 * Public API - TypeDefs
 */
export {
  Template,
  ModelId,
  ModelRequestOptions,
  ModelApi,
  ModelProvider,
  HttpClient,
  HttpClientRequest,
} from "./typeDefs";
