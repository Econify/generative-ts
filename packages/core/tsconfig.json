{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "composite": true,
    "outDir": "./dist",
    "rootDir": "./src",

    "declaration": true,
  },
  "include": ["src/**/*.ts"],
  "exclude": ["src/**/*.spec.ts"],
  "references": []
}
