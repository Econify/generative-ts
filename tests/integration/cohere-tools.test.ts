import { CohereChatApi, createCohereModelProvider } from "@packages/core";

test("Cohere - Chat (Tools)", async () => {
  // arrange
  const cohereChat = createCohereModelProvider({
    api: CohereChatApi,
    modelId: "command-r-plus",
  });

  // act
  // const response = await cohereChat.sendRequest({
  //   prompt: "Will the NY Mets game be a rainout tonight",
  //   preamble: "Answer like Jafar from Aladdin",
  //   tools: [
  //     {
  //       name: "get_weather",
  //       description: "tells you the weather in a certain city",
  //       parameter_definitions: {
  //         city: {
  //           type: "str",
  //           description: "the city you want the weather for",
  //         },
  //       },
  //     },
  //   ],
  // });

  // get back a chat_history that looks like this:
  /*
      "chat_history": [
        {
          "role": "USER",
          "message": "Will the NY Mets game be a rainout tonight"
        },
        {
          "role": "CHATBOT",
          "tool_calls": [
            {
              "name": "get_weather",
              "parameters": {
                "city": "New York City"
              }
            }
          ]
        }
      ],
  */

  const response = await cohereChat.sendRequest({
    prompt: "Will the NY Mets game be a rainout tonight",
    preamble: "Answer like Jafar from Aladdin",
    system: "Use the correct JSON output format",
    examplePairs: [
      {
        user: "Heres an example question demonstrating desired JSON format",
        assistant: "{ _correct_answer: 'Im jafar and here is your answer' }",
      },
    ],
    chat_history: [
      {
        role: "USER",
        message: "Will the NY Mets game be a rainout tonight",
      },
      {
        role: "CHATBOT",
        tool_calls: [
          {
            name: "get_weather",
            parameters: {
              city: "New York City",
            },
          },
        ],
      },
      {
        role: "TOOL",
        tool_results: [
          {
            call: {
              name: "get_weather",
              parameters: {
                city: "New York City",
              },
            },
            outputs: [
              {
                weather: "300 feet of rain expected. evacuate immediately.",
              },
            ],
          },
        ],
      },
    ],
    tool_results: [
      {
        call: {
          name: "get_weather",
          parameters: {
            city: "New York City",
          },
        },
        outputs: [
          {
            weather: "300 feet of rain expected. evacuate immediately.",
          },
        ],
      },
    ],
    tools: [
      {
        name: "get_weather",
        description: "tells you the weather in a certain city",
        parameter_definitions: {
          city: {
            type: "str",
            description: "the city you want the weather for",
          },
        },
      },
    ],
  });

  console.log(JSON.stringify(response.chat_history, null, 2));
  console.log(`Text: "${response.text}"`);

  // assert
  // expect(response).toMatchApiSnapshot();
  expect(response).toBeDefined();
});
