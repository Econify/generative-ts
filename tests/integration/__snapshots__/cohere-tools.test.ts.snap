// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cohere - Chat (Tools) 1`] = `
{
  "chat_history": [
    {
      "message": "string",
      "role": "string",
    },
  ],
  "finish_reason": "string",
  "generation_id": "string",
  "meta": {
    "api_version": {
      "version": "string",
    },
    "billed_units": {
      "input_tokens": "number",
      "output_tokens": "number",
    },
    "tokens": {
      "output_tokens": "number",
    },
  },
  "response_id": "string",
  "text": "string",
  "tool_calls": [
    {
      "name": "string",
      "parameters": {
        "city": "string",
      },
    },
  ],
}
`;
